pkg/manager/member/tidb_member_manager.go

```diff
@@ -283,37 +283,54 @@ func (m *tidbMemberManager) syncTiDBService(tc *v1alpha1.TidbCluster) error {
 		return fmt.Errorf("syncTiDBService: failed to get svc %s for cluster %s/%s, error: %s", newSvc.Name, ns, tc.GetName(), err)
 	}
 	oldSvc := oldSvcTmp.DeepCopy()
+	if newSvc.Annotations == nil {
+		newSvc.Annotations = map[string]string{}
+	}
+	if oldSvc.Annotations == nil {
+		oldSvc.Annotations = map[string]string{}
+	}
+	if newSvc.Labels == nil {
+		newSvc.Labels = map[string]string{}
+	}
+	if oldSvc.Labels == nil {
+		oldSvc.Labels = map[string]string{}
+	}
 	util.RetainManagedFields(newSvc, oldSvc)
 
 	equal, err := controller.ServiceEqual(newSvc, oldSvc)
 	if err != nil {
 		return err
 	}
 	annoEqual := util.IsSubMapOf(newSvc.Annotations, oldSvc.Annotations)
+	labelEqual := util.IsSubMapOf(newSvc.Labels, oldSvc.Labels)
 	isOrphan := metav1.GetControllerOf(oldSvc) == nil
 
-	if !equal || !annoEqual || isOrphan {
-		svc := *oldSvc
-		svc.Spec = newSvc.Spec
-		err = controller.SetServiceLastAppliedConfigAnnotation(&svc)
-		if err != nil {
-			return err
-		}
-		svc.Spec.ClusterIP = oldSvc.Spec.ClusterIP
-		// apply change of annotations if any
-		for k, v := range newSvc.Annotations {
-			svc.Annotations[k] = v
-		}
-		// also override labels when adopt orphan
-		if isOrphan {
-			svc.OwnerReferences = newSvc.OwnerReferences
-			svc.Labels = newSvc.Labels
-		}
-		_, err = m.deps.ServiceControl.UpdateService(tc, &svc)
-		return err
+	if equal && annoEqual && labelEqual && !isOrphan {
+		return nil
 	}
 
-	return nil
+	svc := *oldSvc
+	svc.Spec = newSvc.Spec
+	err = controller.SetServiceLastAppliedConfigAnnotation(&svc)
+	if err != nil {
+		return err
+	}
+	svc.Spec.ClusterIP = oldSvc.Spec.ClusterIP
+	// apply change of annotations if any
+	for k, v := range newSvc.Annotations {
+		svc.Annotations[k] = v
+	}
+	// apply change of labels if any
+	for k, v := range newSvc.Labels {
+		svc.Labels[k] = v
```

Comment:
Why merge labels here instead of overwriting?
The same comment for the annotations.

Reply:
update in https://github.com/pingcap/tidb-operator/pull/3892/commits/163817430a9fec825008aa2a341593415ea6174b